README.md
# BasitApi [![Build Status](https://travis-ci.org/omerucel/basitapi.png)](https://travis-ci.org/omerucel/basitapi)

RESTful API oluşturmanızda size yardımcı olan basit bir kütüphane. Django'nun
sınıf temelli View yapısı üzerine kurulu. Bir RESTful API'de olması gereken
bazı temel özellikleri sunar ve başka da bir işe karışmaz.

Sunduğu bazı temel özellikler:

*   *Bağlantıdaki method parametresi ile HTTP_METHOD başlığının ezilmesi.*

    Bazı istemciler sadece bazı metodları göndermekle kısıtlandırılabilirler. Bu
    türlü bir durumda API'nin doğru çalışabilmesi için bağlantı içinde *method*
    isimli parametreyi kullanabilirsiniz. Bu parametre ile HTTP_METHOD ezilir ve
    gönderdiğiniz method parametresindeki değere göre işlemler yürütülür.

*   *Her zaman 200 HTTP kodu ile dönüş.*

    Bazı istemciler sadece bazı hata kodlarına yanıt verebilirler. BasitApi bu tür durumlar için
    *suppress_response_codes* parametresine destek vermekte. Bu parametre değeri
    1 olarak gönderildiğinde durumu ne olursa olsun tüm yanıtlar 200 kodu ile
    gönderilmekte.

*   *Yanıt formatı.*

    Hem *Accept* başlık bilgisine hem de bağlantı dosya son ekine göre yanıt
    formatı ayarlanabilmekte.


## Kurulum

PyPi üzerinden kurulum için aşağıdaki komutu kullanabilirsiniz:

```
$ pip install basitapi
```

Github üzerinden kurulum için aşağıdaki komutu kullanabilirsiniz:

```
$ pip install -e git://github.com/omerucel/basitapi.git#egg=basitapi
```

## Kullanım

### settings.py

```python
INSTALLED_APPS=(
    ...
    'basitapi'
)
```

### urls.py

Bağlantı dosya son ekine göre yanıt formatını ayarlamak istemiyorsanız
bağlantılarınızı *format_suffix_patterns* ile tekrar formatlamanıza gerek yok!

```python
from basitapi.urlpatterns import format_suffix_patterns
from appname.views import ViewName

urlpatterns = patterns(
    url(r'^foo', ViewName.as_view()),
    ...
)

urlpatterns = format_suffix_patterns(urlpatterns)
```

### views.py

```python

from basitapi.exception import ApiException
from basitapi.response import ApiResponse
from basitapi.views import ApiView

class ViewName(ApiView):
    def get(self, request):
        return ApiResponse({
            'result' : 'data'
        })

    def post(self, request):
        raise ApiException('Bir hata olustu', status=403, application_code=1111)
```


CHANGELOG.txt
=== 0.1.1 ===

* Django 1.5 support

=== 0.1 ===

Initial commit


basitapi/tests/urls.py
from django.conf.urls import patterns, url

from basitapi.urlpatterns import format_suffix_patterns
from basitapi.tests.integration_tests import views_tests

urlpatterns = patterns('',
    url(r'sample$', views_tests.SampleView.as_view()),
    url(r'error$', views_tests.SampleErrorView.as_view())
)

urlpatterns = format_suffix_patterns(urlpatterns)


basitapi/tests/runtests.py
import os
import sys

from django.conf import settings

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.admindocs',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.messages',
    'django.contrib.sessions',
    'django.contrib.staticfiles',
    'django.contrib.sitemaps',
    'django.contrib.sites',
    'basitapi',
    'django_nose',
]


if not settings.configured:
    settings.configure(
        DATABASES={
            'default':{
                'ENGINE': 'django.db.backends.sqlite3',
                'NAME': ':memory:',
            }
        },
        INSTALLED_APPS=INSTALLED_APPS,
        ROOT_URLCONF='basitapi.tests.urls',
    )

from django_nose import NoseTestSuiteRunner

def runtests(*test_args):
    failures = NoseTestSuiteRunner(verbosity=2, interactive=True).run_tests(test_args)
    sys.exit(failures)

if __name__ == '__main__':
    runtests(*sys.argv[1:])


basitapi/tests/factories.py
import factory

from django.contrib.auth.models import User

class UserFactory(factory.Factory):
    FACTORY_FOR = User

    id = factory.Sequence(lambda a: int(a)+1)
    username = factory.Sequence(lambda a: 'username%d' %(int(a)+1))
    password = factory.Sequence(lambda a: 'password%d' %(int(a)+1))

    @classmethod
    def _prepare(cls, create, **kwargs):
        password = kwargs.pop('password', None)
        user = super(UserFactory, cls)._prepare(create, **kwargs)

        if password:
            user.set_password(password)
            if create:
                user.save()

        return user


basitapi/response.py
#-*- coding: utf-8 -*-

"""
Api yanıtı ile ilgili sınıf ve fonksiyonları içerir.
"""

from django.utils import simplejson

class ApiResponse:
    """
    API yanıtlarının taşınması için kullanılır. Yanıtın düzgün bir şekilde 
    istenilen formata çevirir.
    """
    @staticmethod
    def okay(content=None):
        """
        HTTP 200 status code
        """
        return ApiResponse(content, status=200)

    @staticmethod
    def created(content=None):
        """
        HTTP 201 status code
        """
        return ApiResponse(content, status=201)

    @staticmethod
    def accepted(content=None):
        """
        HTTP 202 status code
        """
        return ApiResponse(content, status=202)

    @staticmethod
    def bad_request(content=None):
        """
        HTTP 400 status code
        """
        return ApiResponse(content, status=400)

    @staticmethod
    def unauthorized(content=None):
        """
        HTTP 401 status code
        """
        return ApiResponse(content, status=401)

    @staticmethod
    def payment_required(content=None):
        """
        HTTP 402 status code
        """
        return ApiResponse(content, status=402)

    @staticmethod
    def forbidden(content=None):
        """
        HTTP 403 status code
        """
        return ApiResponse(content, status=403)

    @staticmethod
    def not_found(content=None):
        """
        HTTP 404 status code
        """
        return ApiResponse(content, status=404)

    @staticmethod
    def method_not_allowed(content=None):
        """
        HTTP 405 status code
        """
        return ApiResponse(content, status=405)

    def __init__(self, content=None, status=200):
        if content == None:
            content = {}

        self.content = content
        self.status = status

    def to_json(self):
        """
        Yanıt içeriğini json formatına çevirir.
        """
        return simplejson.dumps(self.content)

    def to_object(self):
        """
        Yanıt içeriğini objeye dönüştürür.
        """
        class Object:
            """
            Temp object
            """
            def __init__(self, kwargs):
                for i in kwargs:
                    setattr(self, i, kwargs[i])

        obj = Object(self.content)
        return obj


basitapi/views.py
#-*- coding: utf-8 -*-

import django

from django.utils import simplejson
from django.views.generic import View
from django.views.decorators.csrf import csrf_exempt
from django.core.context_processors import csrf
from django.contrib.auth.decorators import login_required
from django.contrib.auth import authenticate, login
from django.shortcuts import render_to_response
from django.template import RequestContext
from django.http import HttpResponse, HttpResponseRedirect, QueryDict

#from oauth2app.authorize import Authorizer, MissingRedirectURI, AuthorizationException

from basitapi.exception import ApiException
from basitapi.response import ApiResponse

class ApiView(View):
    """
    View sınıflarına Api desteği kazandırır.
    """

    _format = None

    def http_method_not_allowed(self, request, *args, **kwargs):
        raise ApiException('Method not allowed.',405)

    @csrf_exempt
    def dispatch(self, request, *args, **kwargs):
        """
        Gelen istek application/x-www-form-urlencoded ile gönderilmişse
        raw_post_data içindeki veriler objeye çevrilir ve request.REQUEST güncellenir.
        """

        if django.VERSION[0] >= 1 and django.VERSION[1] > 4:
            request_body = request.body
        else:
            request_body = request.raw_post_data

        if 'application/x-www-form-urlencoded' in request.META.get('CONTENT_TYPE', '') and request.method in ['PUT']:
            request.REQUEST.dicts = (request.POST, request.GET, simplejson.loads(request_body))

        if 'application/json' in request.META.get('CONTENT_TYPE', ''):
            request.REQUEST.dicts = (request.POST, request.GET, simplejson.loads(request_body))

        try:
            # Yanıt formatını belirlemek için önce HTTP_ACCEPT kontrol ediliyor.
            if 'application/json' in request.META.get('HTTP_ACCEPT', ''):
                self._format = 'json'
            else:
                # Bağlantı son ekine göre format belirleniyor.
                self._format = kwargs.get('format', 'json')

                # Destek verilmeyen format isteklerinde hata görüntüle.
                if self._format not in ['json', 'xml']:
                    raise ApiException('Unsupported response format.', 400)

            # format parametrelerden temizleniyor. Boş yere kalabalık olmasın..
            if kwargs.has_key('format'):
                del kwargs['format']

            # method parametresi ile istek yapılırsa ilgili metodun çalıştırılması sağlanıyor. (method=PUT, method=POST vb.)
            # Diğer durumlar için View sınıfının dispatch metodu kullanılır.
            if request.REQUEST.get('method', '').lower() in self.http_method_names:
                handler = getattr(self, request.REQUEST.get('method').lower(), self.http_method_not_allowed)
            else:
                handler = super(ApiView, self).dispatch

            response = handler(request, *args, **kwargs)
        except ApiException, error:
            data = {
                'message' : error.message,
                'status' : error.status
            }

            if not error.application_code == None:
                data.update({'application_code' : error.application_code})

            response = ApiResponse(data, error.status)
        except Exception, error:
            response = ApiResponse({
                'message' : 'Internal Server Error',
                'status' : 500
            }, 500)
        finally:
            # TODO : Yanıtın farklı formatlarda sunulabilmesini destekle
            if isinstance(response, ApiResponse):
                status_code = response.status
                mimetype = 'text/plain'
                if self._format == 'json':
                    mimetype = 'application/json'
                elif self._format == 'xml':
                    mimetype = 'text/xml'

                response = HttpResponse(response.to_json(), '%s; charset=utf-8' %(mimetype))
                response.status_code = status_code

            # İstemci tarafındaki yetersizliklerden dolayı bazen 200 haricindeki hata kodları soruna neden olabiliyor.
            # Bu sorunu aşmak için suppress_response_codes=1 değeri ile istekte bulunmak yeterli.
            if request.GET.get('suppress_response_codes', False) == '1':
                response.status_code = 200

            return response

